driver-core/src/test/functional/com/mongodb/client/model/expressions/ArrayExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+import static com.mongodb.client.model.expressions.Expressions.ofBooleanArray;
+
+@SuppressWarnings({"PointlessBooleanExpression", "ConstantConditions", "Convert2MethodRef"})
+class ArrayExpressionsFunctionalTest extends AbstractExpressionsFunctionalTest {
+    // https://www.mongodb.com/docs/manual/reference/operator/aggregation/#array-expression-operators
+    // (Incomplete)
+
+    private final ArrayExpression<BooleanExpression> arrayTTF = ofBooleanArray(true, true, false);
+
+    @Test
+    public void literalsTest() {
+        assertExpression(Arrays.asList(true, true, false), arrayTTF, "[true, true, false]");
+    }
+
+    @Test
+    public void filterTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/filter/
+        assertExpression(
+                Stream.of(true, true, false)
+                        .filter(v -> v).collect(Collectors.toList()),
+                arrayTTF.filter(v -> v),
+                // MQL:
+                "{'$filter': {'input': [true, true, false], 'cond': '$$this'}}");
+    }
+
+    @Test
+    public void mapTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/map/
+        assertExpression(
+                Stream.of(true, true, false)
+                        .map(v -> !v).collect(Collectors.toList()),
+                arrayTTF.map(v -> v.not()),
+                // MQL:
+                "{'$map': {'input': [true, true, false], 'in': {'$not': '$$this'}}}");
+    }
```

Comment:
I added the following test, which should fail with a ClassCast:

```
        assertExpression(
                Arrays.asList(),
                ofBooleanArray().map((BooleanExpression q) -> {
                    IntegerExpression i = (IntegerExpression) q; // <-- passes
                    return q.not();
                }),
                "{'$map': {'input': [], 'in': {'$not': '$$this'}}}");
```

I don't know how this would be implemented for the general case. It seems that to know the "inner" type for `q` (so that it can be wrapped, to ensure a ClassCast), we need a value, but the array is empty, so there's no value to use. In this example we could infer the value because the constructor is ofBooleanArray. But the general `ofArray<T>` seems like it would need a `BooleanExpression.class` parameter that we could store and switch on. But then I don't know where we could put `BoolEx.class` for something like `ArrayEx<ArrayEx<BoolEx>>`.

Reply:
I'm not sure.  Would have to think more.  But it reminds me of some of the recent work on parameterization of records.